applications/models.py

```diff
@@ -221,28 +221,50 @@ def get_notification_context(self):
 
     def _validate_status(self):
         """
-        If the BerthApplication does not have a BerthLease related to it,
-        it can only have PENDING or EXPIRED statuses.
-
-        With a lease, it can have any of the other statuses.
+        * If the BerthApplication it is a switch application, then the set of available statuses is restricted
+        * If the BerthApplication does not have a BerthLease related to it,
+          it can only have PENDING, EXPIRED or NO_SUITABLE_BERTHS statuses.
         """
-        allowed_statuses_without_lease = [
-            ApplicationStatus.PENDING,
-            ApplicationStatus.EXPIRED,
-            ApplicationStatus.NO_SUITABLE_BERTHS,
-        ]
-
-        if (
-            not hasattr(self, "lease")
-            and self.status not in allowed_statuses_without_lease
-        ):
-            raise ValidationError(
-                _(
-                    "BerthApplication with no lease can only be "
-                    f"{', '.join([status.name for status in allowed_statuses_without_lease])}; "
-                    f"has {self.status.name}"
+        if self.berth_switch:
+            allowed_statuses_for_switch = [
+                ApplicationStatus.PENDING,
+                ApplicationStatus.HANDLED,
+                ApplicationStatus.OFFER_SENT,
+                ApplicationStatus.EXPIRED,
+                ApplicationStatus.REJECTED,
+            ]
+            if self.status not in allowed_statuses_for_switch:
+                raise ValidationError(
+                    _(
+                        "Berth switch application can only be "
+                        f"{', '.join([status.name for status in allowed_statuses_for_switch])}; "
+                        f"has {self.status.name}"
+                    )
+                )
+        else:
+            # not a switch application
+            allowed_statuses_without_lease = [
+                ApplicationStatus.PENDING,
+                ApplicationStatus.EXPIRED,
+                ApplicationStatus.NO_SUITABLE_BERTHS,
+            ]
+
+            if self.berth_switch:
+                # See also the status flow diagram
+                allowed_statuses_without_lease.extend(
+                    [ApplicationStatus.OFFER_SENT, ApplicationStatus.REJECTED]
+                )
```

Comment:
Will it ever reach this case? If `self.berth_switch` it will always go to the first `if` and never fall here.

Reply:
right, fixed